### PR TITLE
Upgrade cryptography, remove pyOpenSSL and ndg-httpsclient

### PR DESCRIPTION
**Pull request by millerdev**
This PR includes a rewrite of SSO certificate handling logic to use _cryptography_ as [recommended by the _pyOpenSSL_ docs](https://www.pyopenssl.org/en/stable/api/crypto.html):

> [pyca/cryptography](https://cryptography.io/) is likely a better choice than using this module.

Which allows us to remove the _pyOpenSSL_ dependency from all except docs-requirements.in. Note that _pyOpenSSL_ depends on _cryptography_, so the _cryptography_ dependency can be moved to base-requirements.in.

_ndg-httpsclient_ and _pyasn1_ were also removed from our requirements.in files since we do not appear to be using _ndg-httpsclient_ anymore and _pyasn1_ is not used directly by our code (it is required by other dependencies). They were originally added in https://github.com/dimagi/commcare-hq/pull/20863, although it's unclear why they were required at that time.

:blowfish: Review by commit.

### Safety story

Changes to `corehq/apps/sso/certificates.py` is covered by automated tests in that app.

### Automated test coverage

Yes.

### QA Plan

If problems arise from this change, it will likely come the service provider certificate we give to Azure AD / Microsoft Entra ID. Could the QA team please go through the full process of re-creating an Azure AD Identity Provider on staging for `@dimagi.org` and then logging in with an SSO user? Thanks!

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/sso/tests/test_backends.py
Comment (jingcheng16) [inline corehq/apps/sso/tests/test_backends.py:27]: Thank you! I'm working on this file and also plan to do the same thing. Now I can just wait for this PR get merged and do a rebase. 😂

File: corehq/apps/sso/tests/test_backends.py
Comment (jingcheng16) [inline corehq/apps/sso/tests/test_backends.py:61]: Do we only delete `WebUser` here? and still keep `Django User`?

Comment (millerdev) [reply corehq/apps/sso/tests/test_backends.py:61]: The Django user will get cleaned up when the test transaction rolls back (because it's in SQL, not Couch).

File: requirements/dev-requirements.txt
Comment (millerdev) [reply requirements/dev-requirements.txt:551]: Yes, it's required for `docs-requirements` and because of that, it's also included in `dev-requirements` because `docs-requirements` is included in `dev-requirements`.

You are correct, my PR title was not quite accurate since we still depend on it for our docs builds.

File: corehq/apps/sso/tests/test_backends.py
Comment (millerdev) [reply corehq/apps/sso/tests/test_backends.py:27]: This has to go through QA, so might be a while before it gets merged. You could cherry-pick the commit that contains that into your PR if you want it sooner.

File: corehq/apps/sso/certificates.py
Comment (AmitPhulera) [inline corehq/apps/sso/certificates.py:15]: Although the changes look okay but are there tests for this file?

Comment (millerdev) [reply corehq/apps/sso/certificates.py:15]: There are tests that exercise the functions in this file (indirectly). I didn't find any unit tests for this file specifically.

File: requirements/dev-requirements.txt
Comment (millerdev) [inline requirements/dev-requirements.txt:3]: @biyeun this change is unexpected. What version of `pip-tools` do you have? Should be 7.3.0

Comment (biyeun) [reply requirements/dev-requirements.txt:3]: ah i had `6.8.0`

**Review by biyeun — APPROVED** [review]

**Pull request merged**
